src/lib/modules/deployment/contract_deployer.js

```diff
@@ -209,32 +209,46 @@ class ContractDeployer {
 
     async.waterfall([
       function doLinking(next) {
+
+        if (!contract.linkReferences || !Object.keys(contract.linkReferences).length) {
+          return next();
+        }
         let contractCode = contract.code;
-        self.events.request('contracts:list', (_err, contracts) => {
-          for (let contractObj of contracts) {
-            let filename = contractObj.filename;
-            let deployedAddress = contractObj.deployedAddress;
-            if (deployedAddress) {
-              deployedAddress = deployedAddress.substr(2);
-            }
-            let linkReference = '__' + filename + ":" + contractObj.className;
-            if (contractCode.indexOf(linkReference.substr(0, 38)) < 0) { // substr to simulate the cut that solc does
-              continue;
-            }
-            if (linkReference.length > 40) {
-              return next(new Error(__("{{linkReference}} is too long, try reducing the path of the contract ({{filename}}) and/or its name {{contractName}}", {linkReference: linkReference, filename: filename, contractName: contractObj.className})));
-            }
-            let toReplace = linkReference + "_".repeat(40 - linkReference.length);
-            if (deployedAddress === undefined) {
-              let libraryName = contractObj.className;
-              return next(new Error(__("{{contractName}} needs {{libraryName}} but an address was not found, did you deploy it or configured an address?", {contractName: contract.className, libraryName: libraryName})));
-            }
-            contractCode = contractCode.replace(new RegExp(toReplace, "g"), deployedAddress);
-          }
-          // saving code changes back to the contract object
+        let offset = 0;
+
+        async.eachLimit(contract.linkReferences, 1, (fileReference, eachCb1) => {
```

Comment:
Do you think it would be possible to do a flat map, that way we avoid loop in loop
Also what do you use `eachLimit` if the limit is always 1?

Reply:
Flat Map only flattens one level, but also I didn't think about it.

As for why using `async.each` even if I limit to 1, it's because it's easier to deal with callbacks and errors with `async` methods. I started only using `array.forEach` but handling errors and early exits was painful.
We could probably lift the limit, but anyway, since js is single threaded, it won't change a thing.

Reply:
I guess the complexity at least for me here come from the 3 loops + the 1 request which make it 
4 levels. I find it always hard to read especially when it ends with lot of `eachCbN`

Hence I proposed to first build a data structure that then allow simple logic. (flatMap was maybe not the right one)
Feel free to ignore

Reply:
Understood. I don't like the depth either, but didn't know a simpler approach right away. Constructing a flat map beforehand could work indeed. I'll try if I have time

Reply:
Maybe [`for await...of`](https://devdocs.io/javascript/statements/for-await...of) could be of use?  You'd need an iterable of promises for all the callbacks on the `"contracts:contract"` event. Given the nesting (multiple `fileReference` per element of `linkReferences`), might be easiest to implement by yielding from within an `async function*` (note the `*`).